mne/decoding/gsoc.py

```diff
@@ -0,0 +1,79 @@
+# Authors: Jean-Remi King <jeanremi.king@gmail.com>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+from .mixin import TransformerMixin
+from ..epochs import _BaseEpochs
+
+
+class EpochsTransformerMixin(TransformerMixin):
+    def __init__(self, n_chan=None):
+        self.n_chan = n_chan
+        self._check_init()
+
+    def fit(self, X, y=None):
+        return self
+
+    def fit_transform(self, X, y=None):
+        self.fit(X, y)
+        return self.transform(X)
+
+    def _reshape(self, X):
+        """Recontruct epochs to get a n_trials * n_chan * n_time"""
+        if isinstance(X, _BaseEpochs):
+            X = X.get_data()
+            # TODO: pick data channels (EEG/MEG/SEEG/ECOG if epochs)
+        elif not isinstance(X, np.ndarray):
+            raise ValueError('X must be an Epochs or a 2D or 3D array, got '
+                             '%s instead' % type(X))
+        elif (X.ndim != 3) or (self.n_chan is None):
+            n_epoch = len(X)
+            n_time = np.prod(X.shape[1:]) // self.n_chan
```

Comment:
It only works of X is in 3D already.

This makes life easier when you pipe multiple epochs transformer since you
only need to initialize the n_chan once, you can vary dimensionality
without needing to keep track of it (relevant for feature selection ); note
that you d need to make vectorize X before passing it to a sklearn
transformer (which expects 2D).



On Thursday, 2 June 2016, Asish Panda <notifications@github.com> wrote:

> In mne/decoding/gsoc.py
> <https://github.com/mne-tools/mne-python/pull/3276#discussion_r65576959>:
>
> > +
> > +    def fit_transform(self, X, y=None):
> > +        self.fit(X, y)
> > +        return self.transform(X)
> > +
> > +    def _reshape(self, X):
> > +        """Recontruct epochs to get a n_trials * n_chan * n_time"""
> > +        if isinstance(X, _BaseEpochs):
> > +            X = X.get_data()
> > +            # TODO: pick data channels (EEG/MEG/SEEG/ECOG if epochs)
> > +        elif not isinstance(X, np.ndarray):
> > +            raise ValueError('X must be an Epochs or a 2D or 3D array, got '
> > +                             '%s instead' % type(X))
> > +        elif (X.ndim != 3) or (self.n_chan is None):
> > +            n_epoch = len(X)
> > +            n_time = np.prod(X.shape[1:]) // self.n_chan
>
> When n_chan is not given, how does this work?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3276/files/2e4fb4110344383b594cf8b37f9531cffb95599c#r65576959>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEp7DAueI9f3r_XOmVU0k0nXPomfa6lnks5qHwpWgaJpZM4Isks1>
> .
>